File: Networking Projects/Sniffer Project/sniffer.py
#!/usr/bin/env python3

'''
sniffer.py -- Python packet sniffer

By Jordan Ansell (jordan.ansell@ecs.vuw.ac.nz) July 2017

Use as-is, modification, and/or inclusion in derivative works is permitted only if the original author is credited.

Requires the pcapy library.
To install: python3 -m pip install pcapy -t ./pcapy/

To run:     python3 sniffer.py packets_file.pcap
Or:         tcpdump -w - | python3 ./sniffer.py -

note: awesome website: https://www.binarytides.com/code-a-packet-sniffer-in-python-with-pcapy-extension/
http://simplestcodings.blogspot.com/2010/10/tcp-header-format.html?q=ip+header+format
http://mars.netanya.ac.il/~unesco/cdrom/booklet/HTML/NETWORKING/node020.html

find each header size, ipv4 is variable
pass data to each constructor, data being passed need to start where the last header finished. Can either manipulate packet data variable, or slice off data in python
a[:3] give first three numbers 

can use switches as well, common processes get factored out i.e.process ipv4 then process tcp, use the if statements in those methods then can just call that for the question, rather than always writing nested code. 
'''

from pcapy import pcapy  

import binascii
import struct
import sys
import ipaddress
import socket

def mac2str(mac_bytes):
    mac_string = binascii.hexlify(mac_bytes).decode('ascii')
    mac_pairs = [i+j for i,j in zip(mac_string[0::2], mac_string[1::2])]
    return ':'.join(mac_pairs)

class PacketHeaderBase:
    ''' Base class for packet headers. '''

    def __init__(self, hdr_format, field_names, data):
        ''' Creates a header object from the packet data given.
            Field names are given as an array of strings.
            For more information on struct format strings see:
              https://docs.python.org/3/library/struct.html '''

        self.data = data
        self.hdr_length = struct.calcsize(hdr_format)
        self.field_values = struct.unpack(
            hdr_format, 
            data[:self.hdr_length])

        pkt_dict = dict(zip(field_names, self.field_values))
        for k, v in pkt_dict.items():
            setattr(self, k, v)

class Ethernet(PacketHeaderBase):
    ''' Ethernet header class. '''

    # _fmt and _fields define the structure of an Ethernet packet
    fmt = '!6s6sH'
    fields = ['Destination_Address', 'Source_Address', 'Protocol'] 

    def __init__(self, data):
        super().__init__(Ethernet.fmt, Ethernet.fields, data)

    def __str__(self):
        return "Ethernet payload {}".format( binascii.hexlify(self.payload) )  # what information needs to be printed?

class IPv4_Datagram(PacketHeaderBase):
    ''' Ethernet header class. '''

    # need to find correct IP datagram size
    fmt = '!BBHHHBBH4s4s'
    fields = ['Version_Header_Length', 'Service_Type', 'Total_Length', 'Identification', 'Flags_Fragment_Offset', 'TTL', 'Protocol', 'Header_Checksum', 'Source_Address', 'Destination_Address'] 

    def __init__(self, data):
        super().__init__(IPv4_Datagram.fmt, IPv4_Datagram.fields, data)

    def __str__(self):
        return "IPv4_Datagram payload {}".format( binascii.hexlify(self.payload) )  # what information needs to be printed?

class IPv6_Datagram(PacketHeaderBase):
    ''' Ethernet header class. '''

    # need to find correct IP datagram size
    fmt = '!LHBB16s16s'
    fields = ['Version_TrafficClass_Flow_Label', 'Payload_Length', 'Next_Header', 'Hop_Limit', 'Source_Address', 'Destination_Address'] 

    def __init__(self, data):
        super().__init__(IPv4_Datagram.fmt, IPv4_Datagram.fields, data)

    def __str__(self):
        return "IPv4_Datagram payload {}".format( binascii.hexlify(self.payload) )  # what information needs to be printed?
    
class TCP(PacketHeaderBase):
    ''' TCP header class. '''

    # _fmt and _fields define the structure of an Ethernet packet
    fmt = '!HHLLBBHHH'
    fields = ['Source_Port', 'Destination_Port', 'Sequence_Number', 'Ack_Number', 'Header_Length', 'Reserved', 'Flags', 'Window', 'Checksum', 'Urgent_Pointer', 'Options', 'Padding'] 

    def __init__(self, data):
        super().__init__(TCP.fmt, TCP.fields, data)

    def __str__(self):
        return "TCP payload {}".format( binascii.hexlify(self.payload) )  # what information needs to be printed?

class UDP(PacketHeaderBase):
    ''' UDP header class. '''

    # _fmt and _fields define the structure of an Ethernet packet
    fmt = '!HHHH'
    fields = ['Source_Port', 'Destination_Port', 'Length', 'Checksum'] 

    def __init__(self, data):
        super().__init__(UDP.fmt, UDP.fields, data)

    def __str__(self):
        return "UDP payload {}".format( binascii.hexlify(self.payload) )  # what information needs to be printed?

class ICMP(PacketHeaderBase):
    ''' ICMP header class. '''
    #Need to check here that the ipv4 or ipv6 will have a difference in the main body, check what the procotol numbers mean. The numbers, Mason, what do they mean? 
    
    # _fmt and _fields define the structure of an Ethernet packet
    fmt = '!BBH'
    fields = ['Message_Type', 'Code', 'Checksum'] 

    def __init__(self, data):
        super().__init__(ICMP.fmt, ICMP.fields, data)

    def __str__(self):
        return "ICMP payload {}".format( binascii.hexlify(self.payload) )  # what information needs to be printed?

class IPv6Exten(PacketHeaderBase):
    ''' IPv6Exten header class. '''

    # _fmt and _fields define the structure of an Ethernet packet
    fmt = '!HHHH'
    fields = ['Source_Port', 'Destination_Port', 'Length', 'Checksum'] 
       
    def __init__(self, data):
        super().__init__(IPv6Exten.fmt, IPv6Exten.fields, data)

    def __str__(self):
        return "IPv6Exten payload {}".format( binascii.hexlify(self.payload) )  # what information needs to be printed?
    
def process_IPv4(datagram):
    print ('    Ether type: IPv4')
    fromAdd = socket.inet_ntoa(datagram.Source_Address)
    toAdd = socket.inet_ntoa(datagram.Destination_Address)
    print('        From: ' + str(fromAdd))
    print('        To: ' + str(toAdd))
        
    
def process_IPv6(datagram):
    print('      Ether type: IPv6')
    #need to use mac2str here to find the correct bytes for the 
    fromAdd = mac2str(datagram.Source_Address)
#    toAdd = socket.inet_ntoa(datagram.Destination_Address)
    toAdd = mac2str(datagram.Destination_Address)

    print('        From: ' + str(fromAdd))
    print('        To: ' + str(toAdd))
    
def process_TCP(datagram, length):
    print('         Protocol: TCP')
    print('         Src Port: ' + str(datagram.Source_Port))
    print('         Destination Port: ' + str(datagram.Destination_Port))
    datalength = len(datagram.data) - length
    print('         Payload: (' + str(datalength) + ') bytes')
    print_data(datagram.data[length:])
        
def process_ICMP(datagram, version, length):
    print('         Protocol: ICMP')
    type = socket.ntohs(datagram.Message_Type)
    code = socket.ntohs(datagram.Code)
    if version == 4:
        if type == 0:
            print ('          Echo Reply')
        elif type == 1 or type == 2:
            print ('          Unassigned/Reserved Type')
        elif type == 3:
            print ('          Destination Unreachable')
            print ('          Code was: ' + str(code))
        elif type == 5:
            print ('          Redirect Message')
            print ('          Code was: ' + str(code))
        elif type == 8: 
            print ('           Echo Request')	
        else:
            print ('           The type of the control message was: ' + str(type))
            print ('           The code of the control message was: ' + str(code))			
    else:
        if type < 128: 
            print ('    Error message contained')
            if type == 1:
                print ('          Destination Unreachable')
                print ('          Code was: ' + str(code))
            elif type == 2:
                print ('          Packet too big')
            elif type == 3:
                print('           Time exceeded with code: ' + str(code))	
            else:
                print('           Error message contained type: ' + str(type))
                print('           The code of the message was: ' + str(code))
        else:
            print('     Informational Message contained')		
            if type == 128:
                print ('          Echo Request')
            elif type == 129:
                print ('          Echo Reply')
            elif type == 130:
                print ('          Multicast Listener Query')
            elif type == 131:
                print ('          Multicast Listener Report')
            elif type == 132:
                print ('          Multicase Listener Done')
            elif type == 133:
                print ('          Router Solicitation')
            elif type == 138:
                print ('          Router Renumbering')
            else:
                print('          Informational message contained type: ' + str(type) + ' with code' + str(code))
    print_data(datagram.data[length:])
                      
def process_UDP(datagram, length):
    print('         Protocol: UDP')
    print('         Src Port: ' + str(datagram.Source_Port))
    print('         Destination Port: ' + str(datagram.Destination_Port))
    datalength = len(datagram.data) - length
    print ('Payload: (' + str(datalength) + ') bytes')
    print_data(datagram.data[length:])

def process_IPv6_Exten(datagram, length):
    print('         Protocol: IPv6 Extension')
    #Need to decide what to print for the IPv^ Extension, also check the length
    datalength = len(data) - length
    print ('Payload: (' + str(datalength) + ') bytes')
    print_data(datagram.data[length:])

def print_data(data):
    print (str(data))
    #hexed = binascii.hexlify(data) 
    hexed = data.hex()
    print ('hex:'  + str(hexed))
#    tex = format(b
   # dehexed = codecs.decode(hexed, "hex")
    text = bytearray.fromhex(str(data)).decode()
    #current getting non-hexadecimal number found in fromhex() arg at position 1 error, need to fix this. 
    #TODO

def process_packet(packet_data):
    ''' Function for processing a single packet '''
    parsed_pkt = dict()

    # print( binascii.hexlify(packet_data) )
    # process the datalink header
    parsed_pkt['ethernet'] = Ethernet(packet_data)
    
    ethheaderlength = 14
    totallength = ethheaderlength
    
    #now print the relevant fields
    protocolNum = socket.ntohs(parsed_pkt['ethernet'].Protocol)
    print(str(protocolNum))
    if protocolNum == 8 :
        
        parsed_pkt['ip4 datagram'] = IPv4_Datagram(parsed_pkt['ethernet'].data[totallength:])
        process_IPv4(parsed_pkt['ip4 datagram'])        
        headerlength = ((parsed_pkt['ip4 datagram'].Version_Header_Length) & 0xF)       
    
        IPProtocol = parsed_pkt['ip4 datagram'].Protocol
        print(str(IPProtocol))
        totallength += headerlength
        
        if IPProtocol == 1:
            icmpheaderLength = 4
            parsed_pkt['ICMP'] = ICMP(parsed_pkt['ip4 datagram'].data[totallength:])
            #CheckHeaderLength
            process_ICMP(parsed_pkt['ICMP'], 4, icmpheaderLength)
            
        elif IPProtocol == 6:
            parsed_pkt['TCP'] = TCP(parsed_pkt['ip4 datagram'].data[totallength:])
            #CheckHeaderLength
            tcpHeaderLength = 20
            process_TCP(parsed_pkt['TCP'], tcpHeaderLength)           
        
        elif IPProtocol == 17:
            #CheckHeaderLength
            parsed_pkt['UDP'] = UDP(parsed_pkt['ip4 datagram'].data[totallength:])
            udpheaderlength = 8
            #totallength += udpheaderlength
            process_UDP(parsed_pkt['UDP'],udpheaderlength)
    
        elif IPProtocol == 41:
            parsed_pkt['IPv6 Exten'] = IPv6Exten(parsed_pkt['ip4 datagram'].data[totallength:])
            process_IPv6_Exten(parsed_pkt['IPv6 Exten'],totallength)
            #TODO
        else :
            print('      Ether Type: IPv4')
            print('      Protocol number: ' + str(IPProtocol) +  ' cannot be handled within this packet sniffer')
     
    elif protocolNum == 34525 :
        ipv6HeadLength = 40
        #check TCP, UDP same as IPv4, also check the ICMP to see what is different 
        parsed_pkt['IP6'] = IPv6_Datagram(parsed_pkt['ethernet'].data[totallength:])
        process_IPv6(parsed_pkt['IP6'])
        protocolNum = socket.ntohs(parsed_pkt['IP6'].Next_Header)	
        totallength += ipv6HeadLength
        
        if protocolNum == 1:
            parsed_pkt['ICMP'] = ICMP(parsed_pkt['IP6'].data[totallength:])
            icmpheaderLength = 4
            process_ICMP(parsed_pkt['ICMP'], 6, icmpheaderLength)
        
        elif IPProtocol == 6:
            parsed_pkt['TCP'] = TCP(parsed_pkt['IP6'].data[totallength:])
            tcpHeaderLength = 20
            process_TCP(parsed_pkt['TCP'],tcpHeaderLength)
    
        elif IPProtocol == 17:
            parsed_pkt['UDP'] = UDP(parsed_pkt['IP6'].data[totallength:])
            udpheaderlength = 8
            process_UDP(parsed_pkt['UDP'],udpheaderlength)

        else:
            print('      Ether Type: IPv6')
            print('      Protocol number: ' + str(IPProtocol) +  ' cannot be handled within this packet sniffer')
            size = len(data) - totallength
            print('      The payload size was ' + str(size))
    else:
        print ('    Ether type: ' + str(protocolNum) + ' cannot be handled within this packet sniffer')
        

def main(pcap_filename):
    ''' Main function, loops over packets in the given file, counting and processing each. '''
    print( "Opening file: '{}'".format(pcap_filename) )
    pcap_reader = pcapy.open_offline( pcap_filename )

    count = 0
    while True:
        meta, data = pcap_reader.next()
        if len(data) < 1:
            break  # no more packets 
        
        count += 1
        print("#### Packet {}:".format(count))
        process_packet( data )

    print("Complete. {} packets processed.".format(count))

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Please supply pcap file argument: python3 sniffer.py packets.pcap")
        exit()

    main(sys.argv[1])

File: Networking Projects/Sniffer Project/sniffer.c
/*
 * sniffer.c
 * Main method provided by David C Harrison (david.harrison@ecs.vuw.ac.nz) July 2015
 * Use as-is, modification, and/or inclusion in derivative works is permitted only if 
 * the original author is credited. 
 *
 * To compile: gcc -o sniffer sniffer.c -l pcap 
 * Code tutorials taken from: 
 * http://yuba.stanford.edu/~casado/pcap/section2.html
 * https://www.tcpdump.org/pcap.html
 * https://www.tcpdump.org/manpages/pcap.3pcap.html
 * included header file descriptions came from sources such as below
 * https://unix.superglobalmegacorp.com/Net2/newsrc/netinet/ip.h.html
 * 
 * And code help for printing data taken heavily from: (full creditation down below and explanation of adaption above relevant method)
 * https://www.binarytides.com/packet-sniffer-code-c-libpcap-linux-sockets/
 * 
 *  To run: tcpdump -s0 -w - | ./sniffer -
 *     Or: ./sniffer <some file captured from tcpdump or wireshark>
 */
 
#include <stdio.h>
#include <pcap.h>     
#include <arpa/inet.h>         //for the ntohs translation part of the assignment, converting addresses
#include <netinet/if_ether.h>  //ethernet header import
#include <netinet/ip.h>        //ipv4 header import
#include <netinet/ip6.h>       //ipv6 header import
#include <netinet/tcp.h>       //tcp header import
#include <netinet/udp.h>       //udp header import
#include <netinet/ip_icmp.h>   //icmp for ipv4 header import
#include <netinet/icmp6.h>     //icmp for ipv6 header import
#include <sys/socket.h>


/* Signatures for all the methods in the program */
void process_IPv4(const u_char *, int);
void process_IPv6(const u_char *, int);
void process_TCP(const u_char *, int);
void process_UDP(const u_char *, int);
void process_ICMP(const u_char *, int);
void process_ICMPv6(const u_char *, int);
void print_Data(const u_char *, int);
void got_packet(u_char *, const struct pcap_pkthdr *, const u_char *);

static int packetNum = 0;  // track how many packets have been processed by the program

/* Method taken from the lab1 handout, credited above */
int main(int argc, char **argv)
{
  if (argc < 2) {
    printf("Must have an argument, either a file name or '-'\n");
    return -1;
  }
  
  pcap_t *handle = pcap_open_offline(argv[1], NULL);
  pcap_loop(handle, 1024*1024, got_packet, NULL);
  pcap_close(handle); 
  
  return 0;
}

/* Method which extracts the ethernet protocol and finds out what the type of the packet is, passing the packet to another method afterwards */
void got_packet(u_char *args, const struct pcap_pkthdr *header, const u_char *packet){
  //Keep track of the number of packets processed and find the header length
  printf("Header Length: %d\n", header->len);
  printf("     Packet #: %d \n\n", ++packetNum);

  //assign the packet to the structure created and imported through an netinet header, using the standard length of a ethernet header
  int eth_Length = 14;
  const struct ether_header *eth;
  eth = (struct ether_header*)(packet);
  int eth_Protocol = (int) eth->ether_type;

  //extract the protocol, print a message that the sniffer does not handle this type or pass to ipv4 or ipv6 header for processing.  
  if(eth_Protocol == 8) {
    printf("     Protocol: IPv4\n");
    process_IPv4(packet, eth_Length);
  }
  else if(eth_Protocol == 34525 ) {
    printf("     Protocol: IPv6\n");
    process_IPv6(packet, eth_Length);
  }
  else {
    printf("This sniffer does not handle this Ethernet Protocol: %d \n\n", eth_Protocol);
  }
}
/*Method which handles the IPv4 header, extracting source and desination hosts and extracting the protocol to know which header
will follow from this one, i.e. TCP / UDP...
*/
void process_IPv4(const u_char * packet, int length){
  //create the ip header based on the struct provided in <netinet/ip.h, assigning from the current point in the packet into the fields >
  const struct ip *ip_header; 
  ip_header = (struct ip*)(packet + length);
  
  //create arrays to store the chars from the addresses (source and dest), using the INET_ADDRSTRLEN which is also used in inet_ntop
  char source_address[INET_ADDRSTRLEN];
  char dest_address[INET_ADDRSTRLEN];
  
  //extract header, get the size and add to the current length to know where the next header should start. Extract the protocol to get the next header type
  unsigned char ip_h_size = ip_header->ip_hl;
  int header_Size = (int) ip_header->ip_hl*4; 
  int protocol = (int) ip_header->ip_p;
  int len = length + header_Size;
  
  //convert the source and destination addresses to text form instead of bit form, then store in the previously created char arrays. then print the array, taking advantage of the fact that a string = an array of chars to print all chars in the string immediately.
  inet_ntop(AF_INET, &(ip_header->ip_src), source_address, INET_ADDRSTRLEN);
  inet_ntop(AF_INET, &(ip_header->ip_dst), dest_address, INET_ADDRSTRLEN);
  printf("         From: %s \n", source_address);
  printf("           To: %s \n", dest_address);
  
  //based on the ip's next protocol, continue to process the data packet or exit early if not handled by sniffer.
  if(protocol == 1) {
	  printf("     Protocol: ICMP \n");
	  process_ICMP(packet, len);
	}
	else if(protocol == 6) {
	  printf("     Protocol: TCP \n");
	  process_TCP(packet, len);
	}
	else if(protocol == 17) {
	  printf("     Protocol: UDP \n");
	  process_UDP(packet, len);
	}
  else {
	  printf("     Protocol: %d is not handled within this sniffer \n", protocol);
  }
}

/* Method which handles the IPv6 header, extracting source and desination hosts and extracting the protocol to know which header
will follow from this one, i.e. TCP / UDP...
*/
void process_IPv6(const u_char * packet, int length ){
  //create a header from the netinet header import, use the general size of an ipv6 header and assign the packet to it with the adjustment for the length
  int ipv6_header_size = 40;
  const struct ip6_hdr *ip_header;
  ip_header = (struct ip6_hdr*)(packet + length);
	
  //extract the source and destination addresses using the inet_ntop format to convert the addresses to understandable english
  char source_address[INET6_ADDRSTRLEN];
  char dest_address[INET6_ADDRSTRLEN];
  inet_ntop(AF_INET6, &(ip_header->ip6_src), source_address, INET6_ADDRSTRLEN);
  inet_ntop(AF_INET6, &(ip_header->ip6_dst), dest_address, INET6_ADDRSTRLEN);
  printf("      From: %s \n", source_address);
  printf("        To: %s \n", dest_address);

  //find the length of the headers and the protocol to know which header follows, following the if statement to find the next method to call
  int protocol = ip_header->ip6_nxt;
  int len = length + ipv6_header_size;

  if(protocol == 1) {
	  printf("     Protocol: ICMP \n");
	  process_ICMPv6(packet, len);
  }
	else if(protocol == 6) {
	  printf("     Protocol: TCP \n");
	  process_TCP(packet, len);
	}
	else if(protocol == 17) {
	  printf("     Protocol: UDP \n");
	  process_UDP(packet, len);
	}
	//process outside of the bounds of my understanding, print out the protocol number and move to the next packet
	else {
	  printf("Protocol: %d is not handled within this sniffer \n\n", protocol);
	}
}


/* Method that processes the TCP part of the header, then prints out the data contained once the header has been broken down 
*/
void process_TCP(const u_char * packet, int length){
  //create the pointer for the header, and assign variable names for easy use later.
  const struct tcphdr* tcp_header;
  unsigned short source_port;
  unsigned short destination_port;
  unsigned char *payload;
	
  tcp_header = (struct tcphdr*)(packet + length);
  int len = length + (tcp_header->doff*4);
  //extract the source and destination ports from the header and interpret using the network to host conversion
  source_port = ntohs(tcp_header->source);
  destination_port = ntohs(tcp_header->dest);
  printf("     Src port: %hu\n", source_port);
  printf("    Dest port: %hu\n", destination_port);

  //assign the payload to a u char* adjusting for length of all headers and print out the data
  payload = (u_char*)(packet + len);
  int payload_length = (int) sizeof(payload);
  printf("      Payload: (%d bytes) \n\n", payload_length);
  print_Data(payload, payload_length);
}

/* Method that processes the UDP part of the header, then prints out the data contained once the header has been broken down 
*/

void process_UDP(const u_char * packet, int length ){
  //create a pointer for a header, variables for use in the method
  const struct udphdr* udp_header;
  unsigned short source_port;
  unsigned short destination_port;
  unsigned char *payload;
  int udp_header_size = 8;

  //assign the data from the payload to the structure, adjusting the length of the header
  udp_header = (struct udphdr*)(packet + length);
  
  //extract and print the source and destination sources from the header using network to host conversion  
  source_port = ntohs(udp_header->source);
  destination_port = ntohs(udp_header->source);
  printf("     Src port: %hu \n", source_port);
  printf("    Dest port: %hu \n", destination_port);


  //get payload data
  int len = length + udp_header_size;
  payload = (u_char*)(packet + len);
  int payload_length = sizeof(payload);
  
  //print out protocol details
  printf("Payload: (%d bytes) \n\n", payload_length);
    
  //print out payload data
  print_Data(payload, payload_length);
}

void process_ICMP(const u_char * packet, int length){
  //create the ICMP header with the structure provided in netinet header, using the length of prior headers 
  const struct icmphdr* icmp_header;
  icmp_header = (struct icmphdr*)(packet + length);
  
  //extract the code and type
  int code = (int) icmp_header->code;
  int type = (int) icmp_header->type;
  unsigned char *payload;
  int icmp_header_size = 8;
	
  //print out common type-code messages in English, or default to the general statement of type and code
  if (type == 0){ 
	  printf("     Echo Reply \n");
	}
	else if (type == 1 || type == 2) {
	  printf("     Unassigned/Reserved Type \n");
	}
	else if (type == 3) {
	  printf("     Destination Unreachable\n");
  	printf("     Code was:  %d \n", code);
	}
	else if (type == 5) {
	  printf("     Redirect Message \n");
	  printf("     Code was: %d \n", code);
	}
	else if (type == 8) { 
	  printf("     Echo Request\n");	
	}
	else{ 
    printf("     The type of the control message was: %d \n", type);
    printf("    The code of the control message was: %d \n", code);			
  }
	 
  int len = length + icmp_header_size;
  //find the payload size using the length of the headers total
  payload = (u_char*)(packet + len);
  int payload_length = sizeof(payload);
    
  //print out protocol details
  printf("Payload: (%d bytes) \n \n", payload_length);
    
  //print out payload data
  print_Data(payload, payload_length);
}

/* Method which processes the ICMPv6, understanding what is in the header before printing out data contained in the packet
*/
void process_ICMPv6(const u_char * packet, int length){
  //create a icmp header using the <netinet>header, assigning the next step of the packet to this structure
  const struct icmp6_hdr* icmp6_header;
  icmp6_header = (struct icmp6_hdr*)(packet + length);
  
  //extract type and code from the header, casting to an int for easy comparisons
  int code = (int)icmp6_header->icmp6_code;
  int type = (int)icmp6_header->icmp6_type;
    
  unsigned char *payload;
  int icmp_header_size = 8;
  //understand code and type, using if statements to check. if the code or type is common, then print out an English translation of what the code means
  //otherwise just print the type of message with the code as well. 
  if (type < 128) {
    printf ("    Error message contained \n");
    if (type == 1) {
      printf ("          Destination Unreachable \n");
      printf ("          Code was: %d \n", code);
    }
    else if(type == 2) {
      printf ("          Packet too big\n");
    }
    else if(type == 3) {
      printf("           Time exceeded with code: %d \n", code);	
    }
    else {
      printf("           Error message contained type: %d \n", type);
      printf("           The code of the message was: %d \n", code);
    }
  }
  else{
    printf("     Informational Message contained\n");		
    if (type == 128) {
      printf("    Echo Request\n");
    }
    else if (type == 129) {
      printf("    Echo Reply\n");
    }
    else if (type == 130) {
      printf("    Multicast Listener Query\n");
    }
    else if (type == 131) {
      printf("    Multicast Listener Report\n");
    }
    else if (type == 132) {
      printf("     Multicast Listener Done \n");
    }
    else if (type == 133) {
      printf("     Router Solicitation\n");
    }
    else if (type == 138) {
      printf("     Router Renumbering \n");
    }
    else {
      printf("     Informational message contained type: %d \n", type);
      printf("     The code of the message is: %d \n", code);
    }
  }
	
  int len = length + icmp_header_size;
  payload = (u_char*)(packet + len);
  int payload_length = sizeof(payload);
  
  //print out protocol details
  printf("Payload: (%d bytes) \n", payload_length);
  printf("\n");
  
  //print out payload data
  print_Data(payload, payload_length);
}


/*The following method is taken from the website:
* https://www.binarytides.com/packet-sniffer-code-c-libpcap-linux-sockets/ 
* the code has been adapted slightly to fit the style of this program, adjusting variable names and removing the printing to a file
* as in this assignment we do not print out to a file. There have also been comments added to show the understanding i have over what this
* code is doing  
*/
void print_Data(const u_char * packet, int length) {
  int i , j;
  //step through every character in the data packet
  for(i = 0 ; i < length ; i++){
	  //if a line of hex printing has been completed, then print a space to denote this
    if( i!=0 && i%16==0){ 
      printf("         ");
	  //for each character in the line of hex printing (16 all up)
      for(j=i-16 ; j<i ; j++){
		  //check to see if the character is a number or a letter by comparing to the char value
        if(packet[j]>=32 && packet[j]<=128){
          printf("%c",(unsigned char)packet[j]);
		}
		//if not a number or letter, then print a dot to denote an unknown character
        else {
		  printf("."); 
        }
	  }
	  //print a new line 
      printf("\n");
    } 
         
	//if the end of a line print a space	 
    if(i%16==0){
	  printf("   ");
	}
	//otherwise print the value contained in the data at array position i
    printf(" %02X",(unsigned int)packet[i]);
    
    //at the end of the payload print the last spaces	
    if( i == length-1) { 
      for(j=0;j<15-i%16;j++) {
        printf("   "); 
      }
             
      printf("         ");
      //from point j up to point i in a hex line, check to see if there is a letter or a number based on the char value and print accordingly.       
      for(j = i-i%16 ; j <= i ; j++){
        if(packet[j]>=32 && packet[j]<=128) {
          printf("%c",(unsigned char)packet[j]);
        }
        else {
	      //dot denotes an unknown non-letter or number character
          printf(".");
        }
      }
	  //print a new line 
      printf("\n" );
    }
  }
}